### PR TITLE
 *: support multiple sources replication for mariadb

### PR DESCRIPTION
**Pull request by GregoryIan**
leave canal still only support single source replication

### CODE REVIEW

**Comment by GregoryIan** [comment]
@WangXiangUSTC @siddontang PTAL

File: mysql/mariadb_gtid.go [hunk 1]
```diff
@@ -57,41 +60,157 @@ func (gtid *MariadbGTID) String() string {
...
+	}
+	for domainID, gtid := range s.Sets {
+		clone.Sets[domainID] = gtid.Clone()
+	}
+
 	return clone
 }
+
+// Equal returns true if two mariadb gtid set is same, otherwise return false
+func (s *MariadbGTIDSet) Equal(o GTIDSet) bool {
```
Comment (WangXiangUSTC) [inline mysql/mariadb_gtid.go]: need add length compare

File: mysql/mariadb_gtid.go [hunk 2]
```diff
@@ -57,41 +60,157 @@ func (gtid *MariadbGTID) String() string {
...
+	if str == "" {
+		return s, nil
 	}
 
-	return gtid.DomainID == other.DomainID && gtid.SequenceNumber >= other.SequenceNumber
+	sp := strings.Split(str, ",")
+
+	//todo, handle redundant same uuid
+	for i := 0; i < len(sp); i++ {
+		set, err := ParseMariadbGTID(sp[i])
```
Comment (WangXiangUSTC) [inline mysql/mariadb_gtid.go]: use s.update(sp[i])？

**Comment by WangXiangUSTC** [comment]
LGTM

**Comment by GregoryIan** [comment]
ping @siddontang

**Comment by siddontang** [comment]
please add test for it.

**Comment by GregoryIan** [comment]
@siddontang test was added, PTAL

**Pull request merged**